### DISCUSSION TITLE
Initial javascript request count is really high

### DISCUSSION THREAD

**Discussion by laneme** [vercel/next.js#45570] [Help] [answered]
Let me show you my waterfall real quick

![image](https://user-images.githubusercontent.com/26094878/216770279-ae2f3237-ca63-4818-ae70-fe6e55550791.png)

After the css (5th request) is loaded there is 16 javascript requests. Then two menifests.

This is blocking some critical assets that wouldve been visible much earlier. I did use image priorities, and few of the assests does load eagerly as expected. But others cant fight the network manager. I tested it multiple times

Why not just load the javascript chunks in fewer requests, instead of bombing my network quota with 10s of <1kb requests T.T

I've briefly seen the announcement of next 13 app dir, and there might've been certain improvements regarding this. But for now i'm kinda stuck because of an ongoing issue https://github.com/vercel/next.js/issues/43878

Is there something I might've done wrong? My website has fairly simple structure. It is not script heavy.
I would like to note that on development, it doesnt happen, even though each chunk is much much bigger, requests are kept to a minimum.

**Comment by icyJoseph** [comment, marked as answer]
I am not sure why your bundle broke into so many parts. Are you using a bunch of NPM packages? Usually they are bundled separately, like if I use `axios` it'll be its own bundle and so on.

If you have critical assets, you might want to preload them. See here: https://github.com/vercel/next.js/discussions/42886#discussioncomment-4155003

Basically, you place an HTML tag at top of your application and that'll queue the asset before the scripts. Note that the scripts are already deferred, so you gotta place the assets in the HTML, before them because the defer queue follows HTML precedence.

You might want to also study your `coverage` in the Chrome Dev Tools. https://developer.chrome.com/docs/devtools/coverage/, perhaps you can drop a package, you think you need? Or load them dynamically via `import()` or `next/dynamic`?

**Reply by laneme** [reply]
Interesting

Preloading is being more of hassle than real benefit in my case, and cannot do them in `_document.tsx` either because that doesnt support ssr and client logic.

I'm curious though, why next bundles each package into their own bundle, if it can clearly bring network overhead. 

Thank you for the advice of using dynamic imports / and coverage tools. It helped a bit. I honestly did not realize next would bundle those tiny packages into multiple bundles.

**Reply by icyJoseph** [reply]
Could you run `@next/bundle-analyzer` and share a screenshot? A large chunk has its problems too, but yeah, separating things into chunks helps when there's multiple entry points to an app, like maybe page A uses axios, but not page B, and so on.

**Reply by laneme** [reply]
Oh my bad, Check out these

![image](https://user-images.githubusercontent.com/26094878/216835248-04d97306-5e29-4037-9804-6df92d9a18e0.png)

**Found these On the bottom right**
![image](https://user-images.githubusercontent.com/26094878/216835261-ab46a611-f32a-4001-924d-d84acc6c459b.png)

9 out of these 11 chunks belongs to `react-icons`. React icons exports the components like these
```ts
import { BsArrowRightShort, BsCheck } from 'react-icons/bs'
import { CgArrowLongRight, CgClose } from 'react-icons/cg'
```
And basically a bundle is being made for each submodule (idk if submodule is the correct term). One for `react-icons/bs`, one for `react-icons/cg` and so on

What seems odd to me is that, `framer-motion` and `swiper`, two entirely different packages being concatenated into one chunk, but idk why not these react-icons. I did not notice this before but it seems one of my older project (next 11) is having the same fate with this package.

How should I go debug this issue? Few hints would be really helpful.

I'll try re-exporting every react-icons from a local library and see if that can hack around this. [Edit, it doesnt :c]

**Reply by icyJoseph** [reply]
Right the thing is that for example if page A uses `foo` and `bar` npm packages, and then page B uses `foo`, `bar` and `icons`, we could make a bundle out of `foo` and `bar`, so that page A and B can download it, and another `icons` bundle which B can download, without incrementing page A's load.

The rules of bundling and how it is implemented are a bit tricky though. It is one selling point of Next.js, that you don't have to think about it, until something like your situation happens.

I've seen this react-icons library before, they have some sort of structure where each icon is almost its own package, in order to prune every other icon that you don't use and thus reduce total bundle size, at the expense of having more little bundles. 

I am not sure if there's a way to control WebPack so that it specifically bundles together every `react-icons` sub-bundle.

**Reply by laneme** [reply]
Owh. Understandable.

I'll take a look at webpack api and see if I can do anything about it.

Thank you for your insights so far.  You're the best <3
